### DISCUSSION TITLE
Using venv for MicroPython

### DISCUSSION THREAD

**Discussion by saraverbeecke** [micropython/micropython#10255] [Using MicroPython] [answered]
Hi,
I'm using the new `mip` tool to install some libs and it works great, but does anyone know if we can create a virtual env like we use to do for standard python: `python -m venv my_virtual_env`?
I don't want to mess up my global unix port of micropython and instead use separate virtual envs for my MicroPython projects...
Any help/tips welcome!
Thanks!

**Comment by dlech** [comment, marked as answer]
I think you can run `mip` with the `--target` option to install packages to a specific directory, then set the `MICROPYPATH` environment variable to that path before running `micropython`.

**Reply by saraverbeecke** [reply]
Thanks! I wasn't aware of this option yet :-)

**Comment by jimmo** [comment]
We provide a micro version of the venv library in MicroPython lib that you can install on the unix port. You can install this via mip. We are considering making it included by default in the unix port.

So:

```
# install venv package globally
$ micropython -m mip install venv
# create a new venv
$ micropython -m venv venv-path
# enter venv (same as in CPython)
$ . venv-path/bin/activate
$ # now inside micropython venv
...
$ deactivate
```

**Reply by saraverbeecke** [reply]
Wow that is even more interesting! Is there some sort of `mip freeze` command to create a `requirements.txt` file?

**Reply by jimmo** [reply]
There isn't.

I am planning the opposite which is a way to consume a requirements.txt file or equivalent (i.e. manifest.py).

**Comment by adrianblakey** [comment]
Is there an issue with this?

Do I need to do something other than setting PYTHONPATH to enable micropython to find the modules in ~/.micropython/lib

```
 micropython -v
MicroPython v1.26.1 on 2025-09-11; darwin [Clang 17.0.0] version
Use Ctrl-D to exit, Ctrl-E for paste mode
>>> 


micropython -m mip install venv
Installing venv (latest) from https://micropython.org/pi/v2 to /Users/ajb/.micropython/lib
Copying: /Users/ajb/.micropython/lib/argparse.mpy
Copying: /Users/ajb/.micropython/lib/requests/__init__.mpy
Copying: /Users/ajb/.micropython/lib/mip/__init__.mpy
Copying: /Users/ajb/.micropython/lib/mip/__main__.mpy
Copying: /Users/ajb/.micropython/lib/venv/__main__.mpy
Done

PYTHONPATH=/Users/ajb/.micropython/lib
micropython -m venv venv-micro
ImportError: no module named 'venv.__main__'

```

**Reply by jonnor** [reply]
`PYTHONPATH` is for CPython. The MicroPython path is affected by `MICROPYPATH`.

**Comment by Josverl** [comment]
Documentation is lacking but with a little help from our helmed friend:


1. install the venv module:
`micropython -m mip install venv`

2. Run this command with the unix port binary:
`micropython -m venv myenv`

3. Activate the virtual environment:
`source myenv/bin/activate`

This sets up environment variables so invoking 'micropython' will use your venv's configuration.
( no paths needed from here on ) 

4. Install packages into the venv:
`micropython -m mip install iperf3`



5. inspect the venv
```
$ tree myenv
myenv
├── bin
│   ├── activate
│   └── micropython
└── lib
    └── iperf3.mpy

2 directories, 3 files
```

**Reply by adrianblakey** [reply]
I tried again on MacOS - still get this issue …

```
micropython -m mip install venv
Installing venv (latest) from https://micropython.org/pi/v2 to /Users/ajb/.micropython/lib
Exists: /Users/ajb/.micropython/lib/argparse.mpy
Exists: /Users/ajb/.micropython/lib/requests/__init__.mpy
Exists: /Users/ajb/.micropython/lib/mip/__init__.mpy
Exists: /Users/ajb/.micropython/lib/mip/__main__.mpy
Exists: /Users/ajb/.micropython/lib/venv/__main__.mpy

micropython -m venv myenv
ImportError: no module named 'venv.__main__’

which micropython
/opt/homebrew/bin/micropython

micropython -v
MicroPython v1.26.1 on 2025-09-11; darwin [Clang 17.0.0] version
Use Ctrl-D to exit, Ctrl-E for paste mode
>>> 

echo $MICROPYPATH $PYTHONPATH
/Users/ajb/.micropython/lib /Users/ajb/.micropython/lib

echo $SHELL
/bin/bash

```

**Reply by Josverl** [reply]
Sorry, no access to a Mac to repro, feel free to send one 😁

**Reply by jonnor** [reply]
You seem to have a file /Users/ajb/.micropython/lib/venv/__main__.mpy - so you need to debug why that does not appear in the import path. 
The import path is accessible as `sys.path` from within (Micro)Python.

**Reply by aivarannamaa** [reply]
What does `sys.implementation` say? Does it have `_mpy` field?
